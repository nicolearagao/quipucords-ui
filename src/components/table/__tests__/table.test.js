import React from 'react';
import { TableVariant } from '@patternfly/react-table';
import { Table } from '../table';

describe('Table Component', () => {
  it('should render a basic component', async () => {
    const props = {
      isHeader: true,
      columnHeaders: ['lorem', 'ipsum', 'dolor', 'sit']
    };

    const component = await shallowComponent(<Table {...props} />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should allow variations in table layout', async () => {
    const props = {
      isHeader: true,
      columnHeaders: ['lorem ipsum'],
      rows: [{ cells: ['dolor'] }, { cells: ['sit'] }]
    };

    const component = await shallowComponent(<Table {...props} />);
    expect(component).toMatchSnapshot('generated rows');

    const componentNoBorders = await component.setProps({
      isBorders: false,
      isHeader: false
    });
    expect(componentNoBorders).toMatchSnapshot('borders and table header removed');

    const componentAriaLabel = await component.setProps({
      ariaLabel: 'lorem ipsum aria-label',
      summary: 'lorem ipsum summary'
    });
    expect(componentAriaLabel).toMatchSnapshot('ariaLabel and summary');

    const componentClassName = await component.setProps({
      className: 'lorem-ipsum-class',
      variant: TableVariant.compact
    });
    expect(componentClassName).toMatchSnapshot('className and variant');
  });

  it('should allow expandable row content', () => {
    const mockOnExpand = jest.fn();
    const props = {
      onExpand: mockOnExpand,
      rows: [
        { cells: ['dolor'], expandedContent: 'dolor sit expandable content', data: { hello: 'world' } },
        { cells: ['sit'] }
      ]
    };

    const component = renderComponent(<Table {...props} />);
    expect(component.find('tr.quipucords-table__tr')).toMatchSnapshot('expandable row content');

    const input = component.find('button');
    component.fireEvent.click(input);

    expect(mockOnExpand.mock.calls).toMatchSnapshot('expand row event');
    expect(component.find('.quipucords-table__td-expand-expanded')).toMatchSnapshot('expanded row');
  });

  it('should allow expandable cell content', () => {
    const mockOnExpand = jest.fn();
    const props = {
      onExpand: mockOnExpand,
      rows: [
        { cells: [{ content: 'dolor', expandedContent: 'dolor sit expandable content' }], data: { hello: 'world' } },
        { cells: ['sit'] }
      ]
    };

    const component = renderComponent(<Table {...props} />);
    expect(component.find('tr.quipucords-table__tr')).toMatchSnapshot('expandable cell content');

    const input = component.find('button');
    component.fireEvent.click(input);

    expect(mockOnExpand.mock.calls).toMatchSnapshot('expand cell event');
    expect(component.find('.quipucords-table__td-expand-expanded')).toMatchSnapshot('expanded cell');
  });

  it('should allow sortable content', () => {
    const mockOnSort = jest.fn();
    const props = {
      isHeader: true,
      onSort: mockOnSort,
      columnHeaders: [{ content: 'lorem ipsum', isSort: true }],
      rows: [{ cells: ['dolor'] }, { cells: ['sit'] }]
    };

    const component = renderComponent(<Table {...props} />);
    expect(component.find('th')).toMatchSnapshot('sortable content');

    const input = component.find('button');
    component.fireEvent.click(input);
    expect(component.find('th')).toMatchSnapshot('sorted column, asc');

    component.fireEvent.click(input);
    expect(component.find('th')).toMatchSnapshot('sorted column, desc');

    expect(mockOnSort.mock.calls).toMatchSnapshot('sort event');
  });

  it('should allow selectable row content', () => {
    const mockOnSelect = jest.fn();
    const props = {
      isHeader: true,
      columnHeaders: ['lorem ipsum'],
      onSelect: mockOnSelect,
      rows: [{ cells: [{ content: 'dolor' }], dataLabel: 'testing' }, { cells: ['sit'] }]
    };

    const component = renderComponent(<Table {...props} />);
    expect(component.find('tbody tr')).toMatchSnapshot('select row content');

    const input = component.find('input[name="checkrow0"]');
    component.fireEvent.click(input, { currentTarget: {}, target: { checked: true }, checked: true });

    expect(mockOnSelect.mock.calls).toMatchSnapshot('select row input');
  });

  it('should pass child components, nodes when there are no rows', async () => {
    const props = {
      isHeader: true,
      columnHeaders: ['lorem ipsum'],
      rows: []
    };

    const component = await shallowComponent(<Table {...props}>Loading...</Table>);
    expect(component).toMatchSnapshot('children');
  });
});
