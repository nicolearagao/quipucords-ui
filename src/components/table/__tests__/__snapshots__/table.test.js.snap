// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should allow expandable cell content: expand cell event 1`] = `
[
  [
    {
      "cellIndex": 0,
      "data": {
        "data": {
          "hello": "world",
        },
      },
      "isExpanded": true,
      "rowIndex": 0,
      "type": "compound",
    },
  ],
]
`;

exports[`Table Component should allow expandable cell content: expandable cell content 1`] = `
<tr
  class="quipucords-table__tr"
  data-ouia-component-id="OUIA-Generated-TableRow-4"
  data-ouia-component-type="PF4/TableRow"
  data-ouia-safe="true"
>
  <td
    class=" quipucords-table__td  pf-c-table__compound-expansion-toggle"
  >
    <button
      aria-expanded="false"
      class="pf-c-table__button"
      id="expand-toggle-undefined-undefined"
      type="button"
    >
      <span
        class="pf-c-table__text"
      >
        dolor
      </span>
    </button>
  </td>
</tr>
`;

exports[`Table Component should allow expandable cell content: expanded cell 1`] = `
<td
  class="quipucords-table__td quipucords-table__td-expand-expanded quipucords-table__td-expand-wrapper"
  colspan="1"
>
  <div
    class="quipucords-table__td-expand-content"
  >
    <div
      class="pf-c-table__expandable-row-content"
    >
      dolor sit expandable content
    </div>
  </div>
</td>
`;

exports[`Table Component should allow expandable row content: expand row event 1`] = `
[
  [
    {
      "cellIndex": -1,
      "data": {
        "data": {
          "hello": "world",
        },
      },
      "isExpanded": true,
      "rowIndex": 0,
      "type": "row",
    },
  ],
]
`;

exports[`Table Component should allow expandable row content: expandable row content 1`] = `
<tr
  class="quipucords-table__tr"
  data-ouia-component-id="OUIA-Generated-TableRow-1"
  data-ouia-component-type="PF4/TableRow"
  data-ouia-safe="true"
>
  <td
    class="quipucords-table__td quipucords-table__td-expand pf-c-table__toggle"
  >
    <button
      aria-disabled="false"
      aria-expanded="false"
      aria-label="Details"
      aria-labelledby="simple-node0 expand-toggle0"
      class="pf-c-button pf-m-plain"
      data-ouia-component-id="OUIA-Generated-Button-plain-1"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe="true"
      id="expand-toggle0"
      type="button"
    >
      <div
        class="pf-c-table__toggle-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
          />
        </svg>
      </div>
    </button>
  </td>
  <td
    class="undefined quipucords-table__td "
  >
    dolor
  </td>
</tr>
`;

exports[`Table Component should allow expandable row content: expanded row 1`] = `
<td
  class="quipucords-table__td quipucords-table__td-expand-expanded quipucords-table__td-expand-wrapper"
  colspan="2"
>
  <div
    class="quipucords-table__td-expand-content"
  >
    <div
      class="pf-c-table__expandable-row-content"
    >
      dolor sit expandable content
    </div>
  </div>
</td>
`;

exports[`Table Component should allow selectable row content: select row content 1`] = `
<tr
  class="quipucords-table__tr"
  data-ouia-component-id="OUIA-Generated-TableRow-11"
  data-ouia-component-type="PF4/TableRow"
  data-ouia-safe="true"
>
  <td
    class="quipucords-table__td quipucords-table__td-select pf-c-table__check"
  >
    <label>
      <input
        aria-label="Select row 0"
        name="checkrow0"
        type="checkbox"
      />
    </label>
  </td>
  <td
    class=" quipucords-table__td "
  >
    dolor
  </td>
</tr>
`;

exports[`Table Component should allow selectable row content: select row input 1`] = `
[
  [
    {
      "data": {
        "dataLabel": "testing",
      },
      "isSelected": true,
      "rowIndex": 0,
      "type": "row",
    },
  ],
]
`;

exports[`Table Component should allow sortable content: sort event 1`] = `
[
  [
    {
      "cellIndex": 0,
      "data": {},
      "direction": "asc",
    },
  ],
  [
    {
      "cellIndex": 0,
      "data": {},
      "direction": "desc",
    },
  ],
]
`;

exports[`Table Component should allow sortable content: sortable content 1`] = `
<th
  aria-sort="none"
  class="quipucords-table__th pf-c-table__sort"
  scope="col"
>
  <button
    class="pf-c-table__button"
    type="button"
  >
    <div
      class="pf-c-table__button-content"
    >
      <span
        class="pf-c-table__text"
      >
        lorem ipsum
      </span>
      <span
        class="pf-c-table__sort-indicator"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 256 512"
          width="1em"
        >
          <path
            d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
          />
        </svg>
      </span>
    </div>
  </button>
</th>
`;

exports[`Table Component should allow sortable content: sorted column, asc 1`] = `
<th
  aria-sort="ascending"
  class="quipucords-table__th pf-c-table__sort pf-m-selected"
  scope="col"
>
  <button
    class="pf-c-table__button"
    type="button"
  >
    <div
      class="pf-c-table__button-content"
    >
      <span
        class="pf-c-table__text"
      >
        lorem ipsum
      </span>
      <span
        class="pf-c-table__sort-indicator"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 256 512"
          width="1em"
        >
          <path
            d="M88 166.059V468c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12V166.059h46.059c21.382 0 32.09-25.851 16.971-40.971l-86.059-86.059c-9.373-9.373-24.569-9.373-33.941 0l-86.059 86.059c-15.119 15.119-4.411 40.971 16.971 40.971H88z"
          />
        </svg>
      </span>
    </div>
  </button>
</th>
`;

exports[`Table Component should allow sortable content: sorted column, desc 1`] = `
<th
  aria-sort="descending"
  class="quipucords-table__th pf-c-table__sort pf-m-selected"
  scope="col"
>
  <button
    class="pf-c-table__button"
    type="button"
  >
    <div
      class="pf-c-table__button-content"
    >
      <span
        class="pf-c-table__text"
      >
        lorem ipsum
      </span>
      <span
        class="pf-c-table__sort-indicator"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 256 512"
          width="1em"
        >
          <path
            d="M168 345.941V44c0-6.627-5.373-12-12-12h-56c-6.627 0-12 5.373-12 12v301.941H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.569 9.373 33.941 0l86.059-86.059c15.119-15.119 4.411-40.971-16.971-40.971H168z"
          />
        </svg>
      </span>
    </div>
  </button>
</th>
`;

exports[`Table Component should allow variations in table layout: ariaLabel and summary 1`] = `
<Grid>
  <GridItem
    span={12}
  >
    <TableComposable
      aria-label="lorem ipsum aria-label"
      borders={true}
      className="quipucords-table "
      isStriped={false}
      summary="lorem ipsum summary"
      variant="compact"
    >
      <Thead>
        <Tr
          className="quipucords-table__tr"
        >
          <Th
            className="quipucords-table__th"
          >
            lorem ipsum
          </Th>
        </Tr>
      </Thead>
      <Tbody>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              dolor
            </Td>
          </Tr>
        </React.Fragment>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              sit
            </Td>
          </Tr>
        </React.Fragment>
      </Tbody>
    </TableComposable>
  </GridItem>
</Grid>
`;

exports[`Table Component should allow variations in table layout: borders and table header removed 1`] = `
<Grid>
  <GridItem
    span={12}
  >
    <TableComposable
      aria-label={null}
      borders={false}
      className="quipucords-table "
      isStriped={false}
      summary={null}
      variant="compact"
    >
      <Tbody>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              dolor
            </Td>
          </Tr>
        </React.Fragment>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              sit
            </Td>
          </Tr>
        </React.Fragment>
      </Tbody>
    </TableComposable>
  </GridItem>
</Grid>
`;

exports[`Table Component should allow variations in table layout: className and variant 1`] = `
<Grid>
  <GridItem
    span={12}
  >
    <TableComposable
      aria-label={null}
      borders={true}
      className="quipucords-table lorem-ipsum-class"
      isStriped={false}
      summary={null}
      variant="compact"
    >
      <Thead>
        <Tr
          className="quipucords-table__tr"
        >
          <Th
            className="quipucords-table__th"
          >
            lorem ipsum
          </Th>
        </Tr>
      </Thead>
      <Tbody>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              dolor
            </Td>
          </Tr>
        </React.Fragment>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              sit
            </Td>
          </Tr>
        </React.Fragment>
      </Tbody>
    </TableComposable>
  </GridItem>
</Grid>
`;

exports[`Table Component should allow variations in table layout: generated rows 1`] = `
<Grid>
  <GridItem
    span={12}
  >
    <TableComposable
      aria-label={null}
      borders={true}
      className="quipucords-table "
      isStriped={false}
      summary={null}
      variant="compact"
    >
      <Thead>
        <Tr
          className="quipucords-table__tr"
        >
          <Th
            className="quipucords-table__th"
          >
            lorem ipsum
          </Th>
        </Tr>
      </Thead>
      <Tbody>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              dolor
            </Td>
          </Tr>
        </React.Fragment>
        <React.Fragment>
          <Tr
            className="quipucords-table__tr"
          >
            <Td
              className="undefined quipucords-table__td "
            >
              sit
            </Td>
          </Tr>
        </React.Fragment>
      </Tbody>
    </TableComposable>
  </GridItem>
</Grid>
`;

exports[`Table Component should pass child components, nodes when there are no rows: children 1`] = `
<Grid>
  <GridItem
    span={12}
  >
    Loading...
  </GridItem>
</Grid>
`;

exports[`Table Component should render a basic component: basic 1`] = `
<Grid>
  <GridItem
    span={12}
  >
    <TableEmpty
      t={[Function]}
    />
  </GridItem>
</Grid>
`;
